internal/utils/utils.go

```diff
@@ -18,34 +18,63 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/log"
 )
 
-func GetScmProviderFromGitRepository(ctx context.Context, k8sClient client.Client, repositoryRef *promoterv1alpha1.GitRepository, obj metav1.Object) (*promoterv1alpha1.ScmProvider, error) {
+func GetScmProviderFromGitRepository(ctx context.Context, k8sClient client.Client, repositoryRef *promoterv1alpha1.GitRepository, obj metav1.Object) (promoterv1alpha1.GenericScmProvider, error) {
 	logger := log.FromContext(ctx)
 
-	var scmProvider promoterv1alpha1.ScmProvider
-	namespace := obj.GetNamespace()
-
-	objectKey := client.ObjectKey{
-		Namespace: namespace,
-		Name:      repositoryRef.Spec.ScmProviderRef.Name,
+	var provider promoterv1alpha1.GenericScmProvider
+	kind := repositoryRef.Spec.ScmProviderRef.Kind
+	if kind == "" {
+		kind = promoterv1alpha1.ScmProviderKind
 	}
-	err := k8sClient.Get(ctx, objectKey, &scmProvider, &client.GetOptions{})
-	if err != nil {
-		if k8serrors.IsNotFound(err) {
-			logger.Info("ScmProvider not found", "namespace", namespace, "name", objectKey.Name)
-			return nil, fmt.Errorf("ScmProvider not found: %w", err)
+
+	switch kind {
+	case promoterv1alpha1.ClusterScmProviderKind:
+		var scmProvider promoterv1alpha1.ClusterScmProvider
+		objectKey := client.ObjectKey{
+			Name: repositoryRef.Spec.ScmProviderRef.Name,
 		}
 
-		logger.Error(err, "failed to get ScmProvider", "namespace", namespace, "name", objectKey.Name)
-		return nil, fmt.Errorf("failed to get ScmProvider: %w", err)
+		err := k8sClient.Get(ctx, objectKey, &scmProvider, &client.GetOptions{})
+		if err != nil {
+			if k8serrors.IsNotFound(err) {
+				logger.Info("ClusterScmProvider not found", "name", objectKey.Name)
+				return nil, fmt.Errorf("ClusterScmProvider not found: %w", err)
+			}
```

Comment:
Can we just drop the special handling of not found? The only advantage is constructing a slightly clearer error message. I'm not sure it's worth the lines of code.